drivers/timer/stm32_lptim_timer.c

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright (c) 2018 Foundries.io Ltd
+ * Copyright (c) 2019 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/stm32_clock_control.h>
+#include <drivers/timer/system_timer.h>
+#include <sys_clock.h>
+
+#include <spinlock.h>
+
+/*
+ * Assumptions and limitations:
+ *
+ * - system clock based on an LPTIM1 instance, clocked by LSI or LSE
+ * - prescaler is set to 1 (LL_LPTIM_PRESCALER_DIV1 in the related register)
+ * - using LPTIM1 AutoReload capability to trig the IRQ (timeout irq)
+ * - when timeout irq occurs the counter is already reset
+ * - the maximum timeout duration is reached with the LPTIM_TIMEBASE value
+ * - with prescaler of 1, the max timeout (LPTIM_TIMEBASE) is 2seconds
+ */
+
+#define LPTIM_CLOCK CONFIG_STM32_LPTIM_CLOCK
+#define LPTIM_TIMEBASE CONFIG_STM32_LPTIM_TIMEBASE
+
+/* nb of LPTIM counter unit per ms */
+#define COUNT_PER_MS (LPTIM_CLOCK / 1000)
+
+/* nb of LPTIM counter unit per kernel tick  */
+#define COUNT_PER_TICK (LPTIM_CLOCK / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+static u32_t last_cnt;
+static u32_t accumulated_lptim_ms;
+
+static struct k_spinlock lock;
+
+static void lptim_irq_handler(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	if ((LL_LPTIM_IsActiveFlag_ARRM(LPTIM1) != 0)
+		&& (LL_LPTIM_IsEnabledIT_ARRM)(LPTIM1) != 0) {
+
+		k_spinlock_key_t key = k_spin_lock(&lock);
+
+		/* LPTIM1 CNT could have already been reset */
+		u32_t cnt_value = LL_LPTIM_GetAutoReload(LPTIM1) + 1
+				+ LL_LPTIM_GetCounter(LPTIM1);
+
+		/* do not set ARR yet, z_clock_announce will do */
+
+		/* a next timeout will be raised */
+		LL_LPTIM_ClearFLAG_ARRM(LPTIM1);
+		/* reset last_cnt for a next call to 'elapsed' */
+
+		k_spin_unlock(&lock, key);
+
+		/* increase the total nb of lptim count
+		 * used in the z_timer_cycle_get_32() function
+		 */
+		accumulated_lptim_ms += ((cnt_value * 1000) / LPTIM_CLOCK);
```

Comment:
Are we confident about the precision ?
couldn't we store accumulated number of counts and only convert to ms when needed (or in a MACRO) ?

Reply:
yes : we changed to store nb of LPTIM counter unit (instead of ms).
converting to ms gives unaccuracy which is "announced" and considered by next the kernel call to 'set timeout()'

Reply:
in anycase the z_timer_cycle_get_32() function expects an u32 value

Reply:
var is now named accumulated_lptim_cnt